File: typescript/creation-patterns/singleton.ts
class Database {
    private static instance: Database;

    private constructor() {}

    public static getInstance(): Database {
        if (!this.instance) {
            this.instance = new Database();
        }

        return this.instance;
    }
}

function clientCode() {
    const db1 = Database.getInstance();
    const db2 = Database.getInstance();

    console.log(db1 === db2); // true, both variables point to the same instance
}

clientCode();

File: rust/src/bin/creation-patterns/singleton.rs
use std::sync::{Arc, OnceLock};

struct Database;

impl Database {
    fn get_instance() -> Arc<Database> {
        static INSTANCE: OnceLock<Arc<Database>> = OnceLock::new();
        INSTANCE.get_or_init(|| Arc::new(Database)).clone()
    }
}

fn main() {
    let instance1 = Database::get_instance();
    let instance2 = Database::get_instance();

    // Both instances should point to the same database instance
    assert!(Arc::ptr_eq(&instance1, &instance2));

    println!("Singleton pattern implemented successfully.");
}
